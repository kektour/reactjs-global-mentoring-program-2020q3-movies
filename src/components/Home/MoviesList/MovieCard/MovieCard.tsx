import React, { useCallback, useMemo } from 'react';
import { useModal } from '../../../../hooks/useModal';
import { Movie } from '../../../../models/movie';
import { RemoveMovie } from '../../../RemoveMovie';
import { UpdateMovie } from '../../../UpdateMovie';
import { MovieActionModal } from './MovieActionModal';
import { useHistory } from 'react-router-dom';

import styles from './MovieCard.module.scss';

type Props = {
  movie: Movie;
};

export const MovieCard: React.FC<Props> = (props) => {
  const { movie } = props;
  const history = useHistory();
  const deleteMovieModal = useModal();
  const updateMovieModal = useModal();
  const handleRootDivClick = useCallback(() => history.push(`/films/${movie.id}`), [movie, history]);
  const genresAsString = useMemo(() => movie.genres.join(', '), [movie]);

  return (
    <div className={styles.root}>
      <div className={styles.movieContent} onClick={handleRootDivClick}>
        <img className={styles.img} src={movie.poster_path} alt="Not Found" />
        <div className={styles.contentContainer}>
          <div className={styles.title}>{movie.title}</div>
          <div className={styles.year}>{movie.release_date}</div>
        </div>
        <div className={styles.genre}>{genresAsString}</div>
      </div>
      <MovieActionModal
        classes={{
          roundButtonRoot: styles.movieActionModalRoundButtonRoot,
        }}
        onEditClick={updateMovieModal.open}
        onDeleteClick={deleteMovieModal.open}
      />
      <RemoveMovie open={deleteMovieModal.isOpen} onClose={deleteMovieModal.close} id={movie.id} />
      <UpdateMovie open={updateMovieModal.isOpen} movie={movie} onClose={updateMovieModal.close} />
    </div>
  );
};
