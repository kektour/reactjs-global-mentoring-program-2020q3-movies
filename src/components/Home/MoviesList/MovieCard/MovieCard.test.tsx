import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import { MovieCard } from './MovieCard';

const movie = {
  id: 337167,
  title: 'Fifty Shades Freed',
  tagline: "Don't miss the climax",
  vote_average: 6.1,
  vote_count: 1195,
  release_date: '2018-02-07',
  poster_path: 'https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg',
  overview:
    'Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins.',
  budget: 55000000,
  revenue: 136906000,
  genres: ['Drama', 'Romance'],
  runtime: 106,
};

describe('MovieCard', () => {
  const spyQuerySelector = jest.fn();

  beforeAll(() => {
    Object.defineProperty(global.document, 'querySelector', { value: spyQuerySelector });
    ReactDOM.createPortal = jest.fn((e) => e) as any;
  });

  it('Should render component correctly', () => {
    const body = document.createElement('body');
    const div = document.createElement('div');
    spyQuerySelector.mockImplementationOnce(() => body);
    spyQuerySelector.mockImplementationOnce(() => div);
    spyQuerySelector.mockImplementationOnce(() => body);
    spyQuerySelector.mockImplementationOnce(() => div);

    const component = render(<MovieCard movie={movie} />);
    expect(component.asFragment()).toMatchSnapshot();
  });
});
