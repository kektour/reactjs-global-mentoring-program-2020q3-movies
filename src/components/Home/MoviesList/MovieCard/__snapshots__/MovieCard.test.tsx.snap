// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard Should render component correctly 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="movieContent"
    >
      <img
        alt="Not Found"
        class="img"
        src="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
      />
      <div
        class="contentContainer"
      >
        <div
          class="title"
        >
          Fifty Shades Freed
        </div>
        <div
          class="year"
        >
          2018-02-07
        </div>
      </div>
      <div
        class="genre"
      >
        Drama, Romance
      </div>
    </div>
    <div>
      <div
        class="root movieActionModalRoundButtonRoot"
      >
        <div
          class="dot"
        />
        <div
          class="dot"
        />
        <div
          class="dot"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
